extension/observer/ecsobserver/README.md

```diff
@@ -0,0 +1,330 @@
+```
+
+### ECS Task Definition based filter
+
+ECS [task definition](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definitions.html) contains one or
+more containers (like Pod in k8s). Long running applications normally uses [service](#ecs-service-name-based-filter).
+while short running (batch) jobs can
+be [created from task definitions directly](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/scheduling_tasks.html)
+.
+
+The `task` definition matches both task definition name and container name (of not empty). Optional config
+like `metrics_path`, `metrics_ports`, `job_name` can override default value.
+
+```yaml
+# Example 1: Matches all the tasks created from task definition that contains memcached in its arn
+arn_pattern: "*memcached.*"
+```
+
+### Docker Label based filter
+
+Docker label can be specified in task definition. Only `port_label` is used when checking if the container should be
+included. Optional config like `metrics_path_label`, `job_name_label` can override default value.
+
+```yaml
+# Example 1: Matches all the container that has label ECS_PROMETHEUS_EXPORTER_PORT_NGINX
+port_label: 'ECS_PROMETHEUS_EXPORTER_PORT_NGINX'
+---
+# Example 2: Override job name based on label MY_APP_JOB_NAME
+port_label: 'ECS_PROMETHEUS_EXPORTER_PORT_MY_APP'
+job_name_label: 'MY_APP_JOB_NAME'
+```
+
+### Notify Prometheus Receiver of discovered targets
+
+There are three ways to notify a receiver
+
+- Use [file based service discovery](#generate-target-file-for-file-based-discovery) in prometheus config and updates
+  the file.
+- Use [receiver creator framework](#receiver-creator-framework) to create a new receiver for new endpoints.
+- Register as a prometheus discovery plugin.
+
+#### Generate target file for file based discovery
+
+- Status: implemented
+
+This is current approach used by cloudwatch-agent and
+also [recommended by prometheus](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#file_sd_config)
+. It's easier to debug and the main drawback is it only works for prometheus. Another minor issue is fsnotify may not
+work properly occasionally and delay the update.
+
+#### Receiver creator framework
+
+- Status: pending
+
+This is a generic approach that creates a new receiver at runtime based on discovered endpoints. The main problem is
+performance issue as described
+in [this issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1395).
+
+#### Register as prometheus discovery plugin
+
+- Status: pending
+
+Because both the collector and prometheus is written in Go, we can call `discover.RegisterConfig` to make it a valid
+config for prometheus (like other in tree plugins like kubernetes). The drawback is the configuration is now under
+prometheus instead of extension and can cause confusion.
+
+## Known issues
+
+- There is no list watch API in ECS (unlike k8s), and we fetch ALL the tasks and filter it locally. If the poll interval
```

Comment:
can we add jitter on ECS API callings for the collectors in the same cluster?

Reply:
I think jitter may not be necessary unless all the collectors (w/ ecs discovery) are started at the same time. We don't deploy it as a daemon service, so chance of this scenario is low. We do have cache on some API results (ECS TaskDefinition, EC2 Instance) so we should be good. Though I plan to emit some metrics/log so we can know it when some users are getting to much API errors.